best-practices-in-kubernetes-security.md

```diff
@@ -0,0 +1,180 @@
+
+## **The Kubernetes Control Plane and its Components**
+
+The Kubernetes Control Plane a.k.a. the Master Node, serves as the brain of the cluster that manages the worker nodes, and it is responsible for ensuring that the system is up and running and is in perfect working order. It serves as the primary entry point for administrators and users to manage the different nodes in the cluster. The control plane comprises the Kubernetes API server, etcd, and several controllers.
+
+The following are the purposes of each of the components of the control plane:
+
+-   etcd - This is the fundamental component of a Kubernetes cluster used to store configuration data such that it is accessible to each node in a Kubernetes cluster.
+
+-   kube-apiserver - This is a REST-based interface through which all communications pass and controls all management and operational functions in Kubernetes.
+
+-   kube-controller-manager - This is a service that is responsible for keeping a close eye on the state of a cluster.
+
+-   kube-scheduler - This component is accountable for scheduling cluster workloads.
+
+The malicious attackers might try to gain access to the Kubernetes API Server and Control Plane. Once they have been hacked, the attackers may compromise the entire cluster by altering the pods. They can deploy new ones or modify or even delete the existing pods from the cluster.
+
+## **Kubernetes Security: Understanding the Challenges**
+
+While Kubernetes is a popular technology for delivering containerized apps, scaling Kubernetes environments is challenging. Each new container increases the attack surface. To be able to effectively deal with Kubernetes security, it is imperative that you’ve complete visibility into every managed container and application request.
+
+Kubernetes is great at orchestration but not as adept at security. For all deployments, it is thus imperative to employ the appropriate deployment architecture and security best practices. In the sections that follow we’ll take a deep dive into some of the Kubernetes security best practices.
+
+### **Securing the Control Plane**
+
+Integrity monitoring on the most critical Kubernetes files is a vital first step to securing the control plane. This way, you will get an instant notification if the settings of your Kubernetes environment are altered in any way. Note that critical files are those that, if hacked or compromised, might cause the whole cluster to fail.
+
+### **Securing the Kubernetes API Server**
+
+To safeguard your Kubernetes API server, ensure that it is on an authorized network and in a private cluster. To restrict unauthorized access to the Kubernetes API server, you should utilize Kubernetes Role-Based Access Control (RBAC).
+
+Considered the Kubernetes control plane gateway, you should allocate a private IP address to the Kubernetes API server to protect it. Access to the Kubernetes API server from any public IP address should be disabled.
```

Comment:
The objective should always be to work with private nodes by disabling public access to prevent attackers from gaining control of the control plane. You should use a load balancer or an API gateway to expose a service on the web and enable the ports you require. The essence is that we should always strive to apply the least-privileged concept and ensure that everything is closed as a preventative measure.